### ISSUE TITLE
Spin off t1Lib?

### ISSUE THREAD

**Issue by behdad** [fonttools/fonttools#1382]
Since it doesn't interact with rest of codebase, and is not touched or improved like the ret, I suggest we spin it offf into its own historical module and remove from fontTools.

**Comment by justvanrossum** [comment]
It's working and is being used. -1 from me. Same with afmLib.py.

**Comment by typemytype** [comment]
the 1st of april is not even close...

**Comment by behdad** [comment]
> It's working and is being used. -1 from me.

That's why I suggested spinning it off, not plain removing.  It shares pretty much nothing with the rest of codebase.

**Comment by justvanrossum** [comment]
So what? It's not causing any harm staying, moving it away will cause people to do unnecessary work.

**Comment by justvanrossum** [comment]
And I should be thankful you don't want to plain remove it? Funny :)

**Comment by behdad** [comment]
No one said be thankful.  I'm just saying I understand it's in use.  Anyway, whatever. :)

**Comment by justvanrossum** [comment]
I'm just a bit flabbergasted that you not needing a particular fonttools component would be reason enough for spinning it off. 

Sure, it's a rarely used part. Sure, it's not maintained as tightly as the rest (but then again, the Type 1 spec hasn't changed in a while). But the fact that it's _there_ as part of fonttools comes in handy time after time. I used it fairly recently for a client that needed legacy Type 1 support.

Actually, t1Lib plays very well with the rest of fonttools, in that is supports the notion of a dict-like glyph set as well as the pen protocol for glyph drawing.

**Comment by behdad** [comment]
> I'm just a bit flabbergasted that you not needing a particular fonttools component would be reason enough for spinning it off.

Where did I offer *that* as my argument?  Don't put words in my mouth.

Guess what, I don't *need* *any* of the code I work on.

I provided a reason from a software maintainership point of view.  And if I thought it's reason enough, I would have gone ahead and done it, not file an issue to discuss.  I'm not even the module maintainer.

> Sure, it's a rarely used part. Sure, it's not maintained as tightly as the rest (but then again, the Type 1 spec hasn't changed in a while). But the fact that it's _there_ as part of fonttools comes in handy time after time. I used it fairly recently for a client that needed legacy Type 1 support.

That's fine.  No need to continue discussion.  I suggested something, you offered counterargument.  @anthrotype will decide and close issue.  You got into unjustified microagression instead.

> Actually, t1Lib plays very well with the rest of fonttools, in that is supports the notion of a dict-like glyph set as well as the pen protocol for glyph drawing.

Fine!

**Comment by justvanrossum** [comment]
I'm sorry for misrepresenting what you said. I just get really grumpy when changes are suggested that will cause other people to have to do unnecessary work.

**Comment by anthrotype** [comment]
i'm -0, so for the sake of 🕊let's just leave t1Lib alone

**Comment by behdad** [comment]
> I just get really grumpy when changes are suggested that will cause other people to have to do unnecessary work.

This.. is also making assumptions.

I've been maintaining widely deployed, system level, software all my professional life.  My C libraries have a stable API policy for life.  I don't just break other people's system for the heck of it.

The change I suggested could be implement without users ever noticing.  fontTools.t1Lib could import * from t1Lib.  pip install fonttools could automatically install t1Lib.  Worst thing that would have happened would be user getting an exception that would say "Please pip install t1tools".  Changing install instruction once in a decade is not too much to bother others with.

The change I proposed has the exact same reasoning that moving ufoLib into fonttools had.  After all, that change had a lot more "cause other people to have to do unnecessary work." effect...

If anything, I'm more interested in moving Type1 code into a separate module and developing a robust Type1 to OpenType converter there.

One more thing. I'm not implying that you (Just) did, but while on this topic, I like to say one more thing: sometimes when @anthrotype or I make or propose a change or decline one, some say "I know Google doesn't need this but..." or "I know Noto doesn't need this but...".  That's diminishing what we do for this community.  More than half of what we do is not needed by Google / Noto.  Percentage might be closer to 80%...  I have fixed bugs for dozens of people in this community, from Microsoft, Adobe, Apple, down to individual designers, and everyone in between.  None of those were needed by Google / Noto.  So, please, keep in mind that what I, and I suppose Cosimo, say, is our professional opinion as Fre Software developer / maintainers, not Google representatives.

Thanks.

**Comment by twardoch** [comment]
I'll also say that the main reason Type 1 will "always" be with us is that Type 1 is required to read and understand for anyone to deal with CFF. 

The CFF spec uses terms that are defined only in Type 1 (actually sometimes you even have to read the PostScript manual). While it is possible to deal with CFF/CFF2 natively, it's much easier to do certain things on the Type 1 level. 

The original usage of AFDKO was that you made a Type 1 and then MakeOTF would convert it to OTF. All those BlueScale/BlueFuzz things, all the CFF private dict — it's all T1. Type 1 is a predecessor if CFF in a similar sense as TrueType is a predecessor to WOFF. 

So it's not "unrelated" to OpenType. And Type 1 is still a supported font format at least on some desktop OSes.

**Comment by twardoch** [comment]
If a restructuring of fontTools were made, I’d be in favor of the following: 

- spin off *all* the "libs" including ttLib into separate packages so that each has a minimal set of requirements
- make fontTools a metapackage that only does imports (for seamless backwards compatibility) and that installs all the spun-off packages via requirements

That would allow some use cases where only ttLib or varLib or whatever else was installed. 

I'm not sure whether this would be useful for anyone or worth the effort. Since the rent trend is actually quite the opposite (ufoLib and designspaceLib and voltLib were added to fontTools), I don't think what I've described above would be of much use.
